File: pymppwriter/reader.py
"""Read an .mpp back into the Project model.

Field-map driven, exactly like the writer: every offset comes from the file's
own `Props` field maps and every flag from its meta bitmaps, so a file saved by
any MPP14-era Project (2010 through the current Microsoft 365 client) reads
correctly — there are no hard-coded record layouts to go stale between
versions.

    from pymppwriter import read_project
    project = read_project("plan.mpp")
    for task in project.tasks:
        print(task.uid, task.name, task.duration_days)

What comes back is the same `Project` the writer consumes, so a file can be
read, edited and written again. Fields the writer does not model (baselines,
costs, timephased data) are not returned.
"""
from __future__ import annotations
import struct
from datetime import datetime
from typing import Dict, List, Optional

from .cfb import load_cfb
from . import blocks as B
from .writer import (PRJ, NATIVE, RSC_NATIVE, ASSN_NATIVE, REL_TYPES, CONSTRAINT_TYPES,
                     TASK_BASELINE_IDS, Baseline, WORK_SCALE,
                     TASK_META_SIZE, TASK_META2_SIZE, REL_META_SIZE, RSC_META_SIZE,
                     ASSN_META_SIZE, TENTHS_PER_DAY, UNITS_CODES, PCT_SCALE,
                     ESTIMATED_FLAG, Project, Task, Relation, Resource, Assignment,
                     decode_rtf_notes)

_REL_TYPE_NAMES = {v: k for k, v in REL_TYPES.items()}
_CONSTRAINT_NAMES = {v: k for k, v in CONSTRAINT_TYPES.items()}
_UNIT_NAMES = {v: k for k, v in UNITS_CODES.items()}


class MppReadError(RuntimeError):
    """The file is not a project we can read."""


class _Klass:
    """One entity class (tasks, resources, …) with its records and field map."""

    def __init__(self, reader, storage: str, meta_size: int, meta2_size: int, props_key: int):
        self.fm, self.bit = reader._class_map(props_key)
        self.recs, self.metas, self.recs2, self.metas2 = [], [], [], []
        try:
            meta = reader._get(f"{storage}/FixedMeta")
            data = reader._get(f"{storage}/FixedData")
        except KeyError:
            return
        self.metas = B.parse_fixed_meta_auto(meta, meta_size)[2]
        self.recs = B.split_fixed_data(data, self.metas)
        try:
            self.metas2 = B.parse_fixed_meta_auto(reader._get(f"{storage}/Fixed2Meta"), meta2_size)[2]
            self.recs2 = B.split_fixed_data(reader._get(f"{storage}/Fixed2Data"), self.metas2)
        except KeyError:
            pass
        self.vtable, self.vdata = {}, b""
        try:
            self.vtable = B.parse_var_meta(reader._get(f"{storage}/VarMeta"))[1]
            self.vdata = reader._get(f"{storage}/Var2Data")
        except KeyError:
            pass

    def value(self, i: int, native_id: int, fmt: str):
        it = self.fm.get(native_id)
        if it is None:
            return None
        src = self.recs[i] if it.block == 0 else (self.recs2[i] if i < len(self.recs2) else b"")
        size = struct.calcsize(fmt)
        if it.offset + size > len(src):
            return None
        return struct.unpack_from(fmt, src, it.offset)[0]

    def timestamp(self, i: int, native_id: int) -> Optional[datetime]:
        it = self.fm.get(native_id)
        if it is None:
            return None
        src = self.recs[i] if it.block == 0 else (self.recs2[i] if i < len(self.recs2) else b"")
        if it.offset + 4 > len(src):
            return None
        return B.decode_timestamp(src, it.offset)

    def flag(self, i: int, native_id: int) -> bool:
        idx = self.bit.get(native_id)
        if idx is None or i >= len(self.metas):
            return False
        meta2 = self.metas2[i] if i < len(self.metas2) else b""
        return bool(B.meta_bit(self.metas[i], meta2, idx))

    def var(self, uid: int, native_id: int) -> Optional[bytes]:
        entry = self.vtable.get(uid, {}).get(native_id)
        return B.read_var(self.vdata, entry) if entry is not None else None

    def text(self, uid: int, native_id: int) -> str:
        raw = self.var(uid, native_id)
        return B.decode_unicode(raw) if raw else ""


class _Reader:
    def __init__(self, path: str):
        self.root = load_cfb(path)
        try:
            _, self.props, _ = B.parse_props(self._get(f"{PRJ}/Props"))
        except KeyError as exc:
            raise MppReadError(
                f"{path!r} has no {PRJ}/Props stream — not an MPP14 project file"
            ) from exc

    def _get(self, path: str) -> bytes:
        node = self.root
        parts = path.split("/")
        for p in parts[:-1]:
            node = node.children[p]
        return node.children[parts[-1]]

    def _class_map(self, props_key: int):
        fm, bit = {}, {}
        if props_key not in self.props:
            return fm, bit
        for i, it in enumerate(B.parse_field_map(self.props[props_key])):
            tid = it.type_value & 0xFFFF
            if it.in_fixed:
                fm.setdefault(tid, it)
            bit.setdefault(tid, i)
        return fm, bit


def _props_text(props: Dict[int, bytes], key: int) -> str:
    raw = props.get(key)
    return B.decode_unicode(raw) if raw else ""


def read_project(path: str) -> Project:
    """Read an .mpp file into a Project.

    Raises MppReadError if the file is not an MPP14 project.
    """
    r = _Reader(path)
    tasks_k = _Klass(r, f"{PRJ}/TBkndTask", TASK_META_SIZE, TASK_META2_SIZE, B.PROPS_TASK_FIELD_MAP)
    rsc_k = _Klass(r, f"{PRJ}/TBkndRsc", RSC_META_SIZE, 53, B.PROPS_RESOURCE_FIELD_MAP)
    assn_k = _Klass(r, f"{PRJ}/TBkndAssn", ASSN_META_SIZE, 53, B.PROPS_ASSIGNMENT_FIELD_MAP)

    tasks: List[Task] = []
    for i, rec in enumerate(tasks_k.recs):
        if len(rec) <= 100:            # stub rows carry no fields
            continue
        uid = tasks_k.value(i, NATIVE["UNIQUE_ID"], "<I")
        if uid is None or uid == 0:    # uid 0 is the project summary row
            continue
        start = tasks_k.timestamp(i, NATIVE["START"])
        finish = tasks_k.timestamp(i, NATIVE["FINISH"])
        if start is None or finish is None:
            continue
        tenths = tasks_k.value(i, NATIVE["DURATION"], "<i") or 0
        units_word = tasks_k.value(i, NATIVE["ACTUAL_DURATION_UNITS"], "<H") or 0
        constraint = _CONSTRAINT_NAMES.get(tasks_k.value(i, NATIVE["CONSTRAINT_TYPE"], "<H"))
        notes_raw = tasks_k.var(uid, NATIVE["NOTES"])
        tasks.append(Task(
            uid=uid,
            name=tasks_k.text(uid, NATIVE["NAME"]),
            start=start,
            finish=finish,
            duration_days=round(tenths / TENTHS_PER_DAY, 4),
            outline_level=tasks_k.value(i, NATIVE["OUTLINE_LEVEL"], "<H") or 1,
            parent_uid=tasks_k.value(i, NATIVE["PARENT_UID"], "<I") or 0,
            duration_units=_UNIT_NAMES.get(units_word & ~ESTIMATED_FLAG, "d"),
            estimated=bool(units_word & ESTIMATED_FLAG),
            notes=decode_rtf_notes(notes_raw) if notes_raw else "",
            wbs=tasks_k.text(uid, NATIVE["WBS"]) or None,
            constraint=constraint if constraint != "ASAP" else None,
            constraint_date=tasks_k.timestamp(i, NATIVE["CONSTRAINT_DATE"]),
            baselines=_read_baselines(tasks_k, uid),
            percent_complete=tasks_k.value(i, NATIVE["PERCENT_COMPLETE"], "<H") or 0,
            priority=tasks_k.value(i, NATIVE["PRIORITY"], "<H") or 500,
            manual=tasks_k.flag(i, NATIVE["MANUALLY_SCHEDULED"]),
        ))

    resources: List[Resource] = []
    for i, rec in enumerate(rsc_k.recs):
        if len(rec) <= 100:
            continue
        uid = rsc_k.value(i, RSC_NATIVE["UNIQUE_ID"], "<I")
        name = rsc_k.text(uid, RSC_NATIVE["NAME"]) if uid is not None else ""
        if not uid or not name:        # uid 0 is Project's unnamed placeholder
            continue
        max_units = rsc_k.value(i, RSC_NATIVE["MAX_UNITS"], "<d")
        resources.append(Resource(
            uid=uid,
            name=name,
            initials=rsc_k.text(uid, RSC_NATIVE["INITIALS"]),
            email=rsc_k.text(uid, RSC_NATIVE["EMAIL_ADDRESS"]),
            max_units=round((max_units or PCT_SCALE) / PCT_SCALE, 4),
        ))

    assignments: List[Assignment] = []
    for i, rec in enumerate(assn_k.recs):
        task_uid = assn_k.value(i, ASSN_NATIVE["TASK_UNIQUE_ID"], "<I")
        rsc_uid = assn_k.value(i, ASSN_NATIVE["RESOURCE_UNIQUE_ID"], "<i")
        if not task_uid or rsc_uid is None or rsc_uid <= 0:
            continue
        units = assn_k.value(i, ASSN_NATIVE["UNITS"], "<d")
        assignments.append(Assignment(
            task_uid=task_uid,
            resource_uid=rsc_uid,
            units=round((units or PCT_SCALE) / PCT_SCALE, 4),
        ))

    relations = _read_relations(r)

    start_raw = r.props.get(B.PROPS_PROJECT_START_DATE)
    project_start = (B.decode_timestamp(start_raw, 0) if start_raw else None) or (
        min((t.start for t in tasks), default=datetime.now())
    )
    return Project(
        title=_props_text(r.props, B.PROPS_TITLE) or "Project",
        start=project_start,
        tasks=tasks,
        relations=relations,
        resources=resources,
        assignments=assignments,
    )


def _read_baselines(tasks_k: _Klass, uid: int) -> Dict[int, Baseline]:
    """Saved baselines for one task, by slot.

    They live in var data — the fixed baseline fields stay empty in files
    Project writes — and a cleared baseline keeps its entries with the dates
    at NA and the numbers at zero, so those are skipped.
    """
    out: Dict[int, Baseline] = {}
    for slot, ids in enumerate(TASK_BASELINE_IDS):
        start = tasks_k.var(uid, ids["start"])
        finish = tasks_k.var(uid, ids["finish"])
        if start is None and finish is None:
            continue
        b = Baseline(
            start=B.decode_timestamp(start, 0) if start else None,
            finish=B.decode_timestamp(finish, 0) if finish else None,
        )
        dur = tasks_k.var(uid, ids["duration"])
        if dur and len(dur) >= 4:
            b.duration_days = round(struct.unpack("<i", dur[:4])[0] / TENTHS_PER_DAY, 4)
        work = tasks_k.var(uid, ids["work"])
        if work and len(work) >= 8:
            b.work_hours = round(struct.unpack("<d", work[:8])[0] / (WORK_SCALE * 600), 4)
        cost = tasks_k.var(uid, ids["cost"])
        if cost and len(cost) >= 8:
            b.cost = struct.unpack("<d", cost[:8])[0]
        if b.start is None and b.finish is None and not b.duration_days and not b.work_hours:
            continue                     # a cleared slot, not a saved one
        out[slot] = b
    return out


def _read_relations(r: _Reader) -> List[Relation]:
    """Dependencies from TBkndCons.

    Only the first three dwords (uid, predecessor, successor) and the type word
    are mapped; the trailer holding lag moved between eras, so it is read the
    way the writer writes it — 2010 files put lag units first, M365 the lag.
    """
    rel_fm, _ = r._class_map(B.PROPS_RELATION_FIELD_MAP)
    is_2010 = rel_fm.get(9) is not None and rel_fm[9].offset == 0
    try:
        metas = B.parse_fixed_meta_auto(r._get(f"{PRJ}/TBkndCons/FixedMeta"), REL_META_SIZE)[2]
        recs = B.split_fixed_data(r._get(f"{PRJ}/TBkndCons/FixedData"), metas)
    except KeyError:
        return []
    out = []
    for rec in recs:
        if len(rec) < 20:
            continue
        _, pred, succ = struct.unpack_from("<III", rec, 0)
        if not pred or not succ:
            continue
        kind = _REL_TYPE_NAMES.get(struct.unpack_from("<H", rec, 12)[0], "FS")
        lag = struct.unpack_from("<i", rec, 16 if is_2010 else 14)[0]
        out.append(Relation(pred, succ, type=kind, lag_days=round(lag / TENTHS_PER_DAY, 4)))
    return out

File: js/src/reader.ts
/**
 * Read an .mpp back into the project model.
 *
 * Field-map driven exactly like the writer: every offset comes from the file's
 * own `Props` field maps and every flag from its meta bitmaps, so a file saved
 * by any MPP14-era Project (2010 through the current Microsoft 365 client)
 * reads correctly — there are no hard-coded record layouts to go stale.
 *
 * What comes back is the same `Project` the writer takes, so a file can be
 * read, edited and written again. Baselines come back on tasks, resources and
 * assignments; fields the writer does not model (costs, timephased data) do
 * not.
 */
import { Storage, readCfb } from "./cfb.ts";
import * as B from "./blocks.ts";
import {
  ASSN_BASELINE_IDS, ASSN_META_SIZE, ASSN_NATIVE, CONSTRAINT_TYPES, ESTIMATED_FLAG, NATIVE,
  PCT_SCALE, PRJ, REL_META_SIZE, REL_TYPES, RSC_BASELINE_IDS, RSC_META_SIZE, RSC_NATIVE,
  TASK_BASELINE_IDS, TASK_META2_SIZE, TASK_META_SIZE,
  TENTHS_PER_DAY, UNITS_CODES, WORK_SCALE, decodeRtfNotes,
  type Assignment, type Baseline, type Project, type Relation, type Resource, type Task,
} from "./model.ts";

/** The file is not a project we can read. */
export class MppReadError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "MppReadError";
  }
}

const dv = (b: Uint8Array) => new DataView(b.buffer, b.byteOffset, b.byteLength);
const invert = (o: Record<string, number>) =>
  new Map(Object.entries(o).map(([k, v]) => [v, k] as const));
const REL_TYPE_NAMES = invert(REL_TYPES);
const CONSTRAINT_NAMES = invert(CONSTRAINT_TYPES);
const UNIT_NAMES = invert(UNITS_CODES);

/** One entity class with its records and its field map. */
class Klass {
  fm: Map<number, B.FieldItem>;
  bit: Map<number, number>;
  recs: Uint8Array[] = [];
  metas: Uint8Array[] = [];
  recs2: Uint8Array[] = [];
  metas2: Uint8Array[] = [];
  table = new Map<number, Map<number, number>>();
  vdata: Uint8Array<ArrayBufferLike> = new Uint8Array(0);

  constructor(reader: Reader, storage: string, metaSize: number, meta2Size: number, propsKey: number) {
    [this.fm, this.bit] = reader.classMap(propsKey);
    const meta = reader.maybe(`${storage}/FixedMeta`);
    const data = reader.maybe(`${storage}/FixedData`);
    if (!meta || !data) return;
    const parsed = B.parseFixedMetaAuto(meta, metaSize);
    this.metas = parsed.items;
    this.recs = B.splitFixedData(data, parsed.items);
    const meta2 = reader.maybe(`${storage}/Fixed2Meta`);
    const data2 = reader.maybe(`${storage}/Fixed2Data`);
    if (meta2 && data2) {
      const parsed2 = B.parseFixedMetaAuto(meta2, meta2Size);
      this.metas2 = parsed2.items;
      this.recs2 = B.splitFixedData(data2, parsed2.items);
    }
    const varMeta = reader.maybe(`${storage}/VarMeta`);
    const varData = reader.maybe(`${storage}/Var2Data`);
    if (varMeta && varData) {
      this.table = B.parseVarMeta(varMeta).table;
      this.vdata = varData;
    }
  }

  private src(i: number, it: B.FieldItem): Uint8Array {
    return it.block === 0 ? (this.recs[i] ?? new Uint8Array(0)) : (this.recs2[i] ?? new Uint8Array(0));
  }

  value(i: number, fieldId: number, kind: "u32" | "i32" | "u16" | "f64"): number | null {
    const it = this.fm.get(fieldId);
    if (!it) return null;
    const src = this.src(i, it);
    const size = kind === "f64" ? 8 : kind === "u16" ? 2 : 4;
    if (it.offset + size > src.length) return null;
    const d = dv(src);
    if (kind === "u32") return d.getUint32(it.offset, true);
    if (kind === "i32") return d.getInt32(it.offset, true);
    if (kind === "u16") return d.getUint16(it.offset, true);
    return d.getFloat64(it.offset, true);
  }

  timestamp(i: number, fieldId: number): Date | null {
    const it = this.fm.get(fieldId);
    if (!it) return null;
    const src = this.src(i, it);
    if (it.offset + 4 > src.length) return null;
    return B.decodeTimestamp(src, it.offset);
  }

  flag(i: number, fieldId: number): boolean {
    const idx = this.bit.get(fieldId);
    if (idx === undefined || i >= this.metas.length) return false;
    return Boolean(B.metaBit(this.metas[i]!, this.metas2[i] ?? new Uint8Array(0), idx));
  }

  var(uid: number, fieldId: number): Uint8Array | null {
    const off = this.table.get(uid)?.get(fieldId);
    return off === undefined ? null : B.readVar(this.vdata, off);
  }

  text(uid: number, fieldId: number): string {
    const raw = this.var(uid, fieldId);
    return raw ? B.decodeUnicode(raw) : "";
  }
}

class Reader {
  root: Storage;
  props: Map<number, Uint8Array>;

  constructor(bytes: Uint8Array) {
    this.root = readCfb(bytes);
    const raw = this.root.get(`${PRJ}/Props`);
    if (!raw) throw new MppReadError(`no ${PRJ}/Props stream — not an MPP14 project file`);
    this.props = B.parseProps(raw).values;
  }

  maybe(path: string): Uint8Array | undefined {
    return this.root.get(path);
  }

  classMap(propsKey: number): [Map<number, B.FieldItem>, Map<number, number>] {
    const fm = new Map<number, B.FieldItem>();
    const bit = new Map<number, number>();
    const raw = this.props.get(propsKey);
    if (!raw) return [fm, bit];
    B.parseFieldMap(raw).forEach((it, i) => {
      if (it.inFixed && !fm.has(it.fieldId)) fm.set(it.fieldId, it);
      if (!bit.has(it.fieldId)) bit.set(it.fieldId, i);
    });
    return [fm, bit];
  }
}

const round4 = (n: number) => Math.round(n * 10000) / 10000;

/** Read an .mpp into a Project. Throws MppReadError when it is not one. */
export function readProject(bytes: Uint8Array): Project {
  const r = new Reader(bytes);
  const tasksK = new Klass(r, `${PRJ}/TBkndTask`, TASK_META_SIZE, TASK_META2_SIZE, B.PROPS_TASK_FIELD_MAP);
  const rscK = new Klass(r, `${PRJ}/TBkndRsc`, RSC_META_SIZE, 53, B.PROPS_RESOURCE_FIELD_MAP);
  const assnK = new Klass(r, `${PRJ}/TBkndAssn`, ASSN_META_SIZE, 53, B.PROPS_ASSIGNMENT_FIELD_MAP);

  const tasks: Task[] = [];
  tasksK.recs.forEach((rec, i) => {
    if (rec.length <= 100) return; // stub rows carry no fields
    const uid = tasksK.value(i, NATIVE["UNIQUE_ID"]!, "u32");
    if (uid === null || uid === 0) return; // uid 0 is the project summary row
    const start = tasksK.timestamp(i, NATIVE["START"]!);
    const finish = tasksK.timestamp(i, NATIVE["FINISH"]!);
    if (!start || !finish) return;
    const tenths = tasksK.value(i, NATIVE["DURATION"]!, "i32") ?? 0;
    const unitsWord = tasksK.value(i, NATIVE["ACTUAL_DURATION_UNITS"]!, "u16") ?? 0;
    const constraintCode = tasksK.value(i, NATIVE["CONSTRAINT_TYPE"]!, "u16");
    const constraint = constraintCode === null ? undefined : CONSTRAINT_NAMES.get(constraintCode);
    const notesRaw = tasksK.var(uid, NATIVE["NOTES"]!);
    const wbs = tasksK.text(uid, NATIVE["WBS"]!);
    const task: Task = {
      uid,
      name: tasksK.text(uid, NATIVE["NAME"]!),
      start,
      finish,
      durationDays: round4(tenths / TENTHS_PER_DAY),
      outlineLevel: tasksK.value(i, NATIVE["OUTLINE_LEVEL"]!, "u16") || 1,
      parentUid: tasksK.value(i, NATIVE["PARENT_UID"]!, "u32") ?? 0,
      durationUnits: UNIT_NAMES.get(unitsWord & ~ESTIMATED_FLAG) ?? "d",
      estimated: Boolean(unitsWord & ESTIMATED_FLAG),
      notes: notesRaw ? decodeRtfNotes(notesRaw) : "",
      percentComplete: tasksK.value(i, NATIVE["PERCENT_COMPLETE"]!, "u16") ?? 0,
      priority: tasksK.value(i, NATIVE["PRIORITY"]!, "u16") || 500,
      manual: tasksK.flag(i, NATIVE["MANUALLY_SCHEDULED"]!),
    };
    const baselines = readTaskBaselines(tasksK, uid);
    if (Object.keys(baselines).length) task.baselines = baselines;
    if (wbs) task.wbs = wbs;
    if (constraint && constraint !== "ASAP") task.constraint = constraint;
    const cdate = tasksK.timestamp(i, NATIVE["CONSTRAINT_DATE"]!);
    if (cdate) task.constraintDate = cdate;
    tasks.push(task);
  });

  const resources: Resource[] = [];
  rscK.recs.forEach((rec, i) => {
    if (rec.length <= 100) return;
    const uid = rscK.value(i, RSC_NATIVE["UNIQUE_ID"]!, "u32");
    if (!uid) return; // uid 0 is Project's unnamed placeholder
    const name = rscK.text(uid, RSC_NATIVE["NAME"]!);
    if (!name) return;
    const maxUnits = rscK.value(i, RSC_NATIVE["MAX_UNITS"]!, "f64");
    const res: Resource = {
      uid,
      name,
      initials: rscK.text(uid, RSC_NATIVE["INITIALS"]!),
      email: rscK.text(uid, RSC_NATIVE["EMAIL_ADDRESS"]!),
      maxUnits: round4((maxUnits ?? PCT_SCALE) / PCT_SCALE),
    };
    const rbl = readRscBaselines(rscK, uid);
    if (Object.keys(rbl).length) res.baselines = rbl;
    resources.push(res);
  });

  const assignments: Assignment[] = [];
  assnK.recs.forEach((_, i) => {
    const taskUid = assnK.value(i, ASSN_NATIVE["TASK_UNIQUE_ID"]!, "u32");
    const rscUid = assnK.value(i, ASSN_NATIVE["RESOURCE_UNIQUE_ID"]!, "i32");
    if (!taskUid || rscUid === null || rscUid <= 0) return;
    const units = assnK.value(i, ASSN_NATIVE["UNITS"]!, "f64");
    const asn: Assignment = {
      taskUid,
      resourceUid: rscUid,
      units: round4((units ?? PCT_SCALE) / PCT_SCALE),
    };
    const auid = assnK.value(i, ASSN_NATIVE["UNIQUE_ID"]!, "u32");
    const abl = auid ? readAssnBaselines(assnK, auid) : {};
    if (Object.keys(abl).length) asn.baselines = abl;
    assignments.push(asn);
  });

  const relations = readRelations(r);
  const startRaw = r.props.get(B.PROPS_PROJECT_START_DATE);
  const projectStart =
    (startRaw ? B.decodeTimestamp(startRaw, 0) : null) ??
    (tasks.length ? new Date(Math.min(...tasks.map((t) => t.start.getTime()))) : new Date());
  const titleRaw = r.props.get(B.PROPS_TITLE);

  return {
    title: (titleRaw ? B.decodeUnicode(titleRaw) : "") || "Project",
    start: projectStart,
    tasks,
    relations,
    resources,
    assignments,
  };
}

/** One little-endian double, or null when the entry is absent or unset. */
function readDouble(raw: Uint8Array | null): number | null {
  if (!raw || raw.length < 8) return null;
  const value = dv(raw).getFloat64(0, true);
  return value === -1e-6 ? null : value; // Project's "no value" double
}

/**
 * Saved baselines for one task, by slot.
 *
 * They live in var data — the fixed baseline fields stay empty in files
 * Project writes — and a cleared baseline keeps its entries with the dates at
 * NA and the numbers at zero, so those are skipped.
 */
function readTaskBaselines(k: Klass, uid: number): Record<number, Baseline> {
  const out: Record<number, Baseline> = {};
  TASK_BASELINE_IDS.forEach((ids, slot) => {
    const startRaw = k.var(uid, ids.start);
    const finishRaw = k.var(uid, ids.finish);
    if (!startRaw && !finishRaw) return;
    const b: Baseline = {};
    const start = startRaw ? B.decodeTimestamp(startRaw, 0) : null;
    const finish = finishRaw ? B.decodeTimestamp(finishRaw, 0) : null;
    if (start) b.start = start;
    if (finish) b.finish = finish;
    const dur = k.var(uid, ids.duration);
    b.durationDays = dur && dur.length >= 4 ? round4(dv(dur).getInt32(0, true) / TENTHS_PER_DAY) : 0;
    b.workHours = round4((readDouble(k.var(uid, ids.work)) ?? 0) / (WORK_SCALE * 600));
    b.cost = readDouble(k.var(uid, ids.cost)) ?? 0;
    // a cleared slot, not a saved one
    if (!b.start && !b.finish && !b.durationDays && !b.workHours) return;
    out[slot] = b;
  });
  return out;
}

/**
 * Saved baselines for one resource, by slot — work and cost only.
 *
 * Project writes no baseline start or finish on a resource, and clearing a
 * slot leaves the two numbers at zero rather than removing them. A resource
 * with nothing assigned therefore looks exactly like a cleared slot, so an
 * all-zero entry is not reported; that ambiguity is in the format.
 */
function readRscBaselines(k: Klass, uid: number): Record<number, Baseline> {
  const out: Record<number, Baseline> = {};
  RSC_BASELINE_IDS.forEach((ids, slot) => {
    const work = readDouble(k.var(uid, ids.work));
    const cost = readDouble(k.var(uid, ids.cost));
    if (!work && !cost) return;
    out[slot] = { workHours: round4((work ?? 0) / (WORK_SCALE * 600)), cost: cost ?? 0 };
  });
  return out;
}

/**
 * Saved baselines for one assignment, by slot — no duration. A cleared slot
 * keeps its entries with the dates at NA and the work at zero, the same
 * convention tasks use.
 */
function readAssnBaselines(k: Klass, uid: number): Record<number, Baseline> {
  const out: Record<number, Baseline> = {};
  ASSN_BASELINE_IDS.forEach((ids, slot) => {
    const startRaw = k.var(uid, ids.start);
    const finishRaw = k.var(uid, ids.finish);
    if (!startRaw && !finishRaw) return;
    const b: Baseline = {};
    const start = startRaw ? B.decodeTimestamp(startRaw, 0) : null;
    const finish = finishRaw ? B.decodeTimestamp(finishRaw, 0) : null;
    if (start) b.start = start;
    if (finish) b.finish = finish;
    b.workHours = round4((readDouble(k.var(uid, ids.work)) ?? 0) / (WORK_SCALE * 600));
    b.cost = readDouble(k.var(uid, ids.cost)) ?? 0;
    if (!b.start && !b.finish && !b.workHours) return; // a cleared slot
    out[slot] = b;
  });
  return out;
}

/**
 * Dependencies from TBkndCons. Only the first three dwords (uid, predecessor,
 * successor) and the type word are mapped; the trailer holding lag moved
 * between eras, so it is read the way the writer writes it — 2010 files put
 * the lag units first, M365 the lag.
 */
function readRelations(r: Reader): Relation[] {
  const [relFm] = r.classMap(B.PROPS_RELATION_FIELD_MAP);
  const is2010 = relFm.get(9)?.offset === 0;
  const meta = r.maybe(`${PRJ}/TBkndCons/FixedMeta`);
  const data = r.maybe(`${PRJ}/TBkndCons/FixedData`);
  if (!meta || !data) return [];
  const items = B.parseFixedMetaAuto(meta, REL_META_SIZE).items;
  const out: Relation[] = [];
  for (const rec of B.splitFixedData(data, items)) {
    if (rec.length < 20) continue;
    const d = dv(rec);
    const pred = d.getUint32(4, true);
    const succ = d.getUint32(8, true);
    if (!pred || !succ) continue;
    const kind = (REL_TYPE_NAMES.get(d.getUint16(12, true)) ?? "FS") as NonNullable<Relation["type"]>;
    const lag = d.getInt32(is2010 ? 16 : 14, true);
    out.push({
      predUid: pred,
      succUid: succ,
      type: kind,
      lagDays: round4(lag / TENTHS_PER_DAY),
    });
  }
  return out;
}
